### PR TITLE
Validate llm_config passed to ConversableAgent (issue #1522)

### LINKED ISSUE #1522: Confusing error message when `llm_config` is not initialized or is `None`.
### Is your feature request related to a problem? Please describe.

Right now, if an agent's `llm_config` is set to None (default in `ConversableAgent` and `AssistantAgent`), or if `config_list` is empty. 

```python
from autogen import UserProxyAgent, AssistantAgent

# assistant = AssistantAgent("assistant", llm_config={"config_list": []})
assistant = AssistantAgent("assistant")
proxy = UserProxyAgent("user")

proxy.initiate_chat(assistant)
```

We get confusing error message from the openai client:  

```
TypeError: Missing required arguments; Expected either ('messages' and 'model') or ('messages', 'model' and 'stream') arguments to be given
```

### Describe the solution you'd like

We should raise an error early if the `llm_config` is None or `config_list` is empty, as it is guaranteed to lead to error as the `model` is not specified. 

Alternatively, we should set `llm_config` by default to false in `ConversableAgent` and make `llm_config` a required argument in `AssistantAgent`. As the current default `None` is always going to cause error. 

### Additional context

Related:

#1210

### PR DESCRIPTION
**Pull request by gunnarku**
Based on #1522 (and a follow-up discussion in this PR), this commit implements the additional validation checks in `ConversableAgent.`

Add the following validation and `raise ValueError` if:

- The `llm_config` is `None` (validated in `ConversableAgent`).
- The `llm_config` has no `model` specified _and_ `config_list` is empty (validated in `OpenAIWrapper`).
- The `config_list` has at least one entry, but not _all_ the entries have the `model` is specified (validated in `OpenAIWrapper`).

The rest of the changes are code churn to adjust or add the test cases.

## Why are these changes needed?

The root cause is #1522. The intent here is to enforce the contract for input parameters that are passed to `ConversableAgent.` With these changes the following happens:

```python
from autogen import UserProxyAgent, AssistantAgent

# Throws an error because "config_list" is empty and "llm_config" does not
# have a value for "model" specified.
assistant = AssistantAgent("assistant", llm_config={"config_list": []})

# Throws an error because "config_list" has no "model" specified.
assistant = AssistantAgent("assistant", llm_config={"config_list":
                                                    [{"model":"", "api_key":
                                                      "sk-mysecretkey"}]})

# Works.
assistant = AssistantAgent("assistant", llm_config={"config_list": 
                                                   [{"model":"gpt-4", "api_key": "sk-mysecretkey"}]})

# Throws an error because one of the "config_list" entries has an empty model.
assistant = AssistantAgent("assistant", llm_config={"config_list":
                                                    [{"model":"gpt-4",
                                                      "api_key": "sk-mysecretkey"},
                                                     {"model":"", "api_key":
                                                      "sk-mysecretkey"}]})

# Throws an error because no "llm_config" is specified.
assistant = AssistantAgent("assistant")

# Works.
assistant = AssistantAgent("assistant", llm_config=False)

# Works: "llm_config" specifies the "model" attribute.
assistant = AssistantAgent("assistant", llm_config={"model": "gpt-4"})
proxy = UserProxyAgent("user")

proxy.initiate_chat(assistant)
```

## Related issue number

Hopefully fixes #1522.

## Checks

- [ ] I've included any doc changes needed for https://microsoft.github.io/autogen/. See https://microsoft.github.io/autogen/docs/Contribute#documentation to build and test documentation locally.
- [ X] I've added tests (if relevant) corresponding to the changes introduced in this PR.
- [ X] I've made sure all auto checks have passed.

### CODE REVIEW

File: autogen/agentchat/conversable_agent.py [hunk 1]
```diff
@@ -136,6 +136,24 @@ def __init__(
...
+                raise ValueError("Please specify at least one configuration in 'llm_config'.")
+
+            # We know that there's at least one entry in the configuration.
+            # Verify that model is specified as well.
+            model = None
+            if "model" in llm_config["config_list"][0]:
+                model = llm_config["config_list"][0]["model"]
+            if model is None or len(model) == 0:
+                raise ValueError("Please specify a value for the 'model' in 'llm_config'.")
+
```
Comment (sonichi) [inline autogen/agentchat/conversable_agent.py:156]: First, let's do the check for `self.llm_config` before line 164 -- that's the line where the error can happen.
Second, we don't require "config_list" to be given or non-empty.
Third, only if model is not in "llm_config", we require "model" to be in each element of `config_list`.

Comment (gunnarku) [reply autogen/agentchat/conversable_agent.py:156]: @sonichi : Thank you, this feedback makes the validation logic very clear. I'll churn on the code tomorrow ;-)

Comment (ekzhu) [reply autogen/agentchat/conversable_agent.py:156]: If you look at `OpenAIWrapper.

Comment (gunnarku) [reply autogen/agentchat/conversable_agent.py]: @ekzhu and @sonichi : thank you for your guidance. Updated the PR description and code with the updated error handling logic.

File: autogen/oai/client.py [hunk 2]
```diff
@@ -345,15 +345,27 @@ def __init__(self, *, config_list: Optional[List[Dict[str, Any]]] = None, **base
...
         if config_list:
             config_list = [config.copy() for config in config_list]  # make a copy before modifying
             for config in config_list:
+                # We require that each element of `config_list` has a non-empty value
+                # for `model` specified.
+                model = None
+                if "model" in config:
+                    model = config["model"]
+                if model is None or len(model) == 0:
+                    raise ValueError("Please specify a value for the 'model' in 'config_list'.")
```
Comment (ekzhu) [inline autogen/oai/client.py:368]:
```suggestion
                    raise ValueError("Please specify a 'model' value for every item in 'config_list'.")
```

File: autogen/agentchat/conversable_agent.py
Comment (ekzhu) [inline autogen/agentchat/conversable_agent.py:1]: @sonichi Right now the only valid values for `llm_config` is `False` and a dictionary. However, the default argument for `llm_config` in `ConversableAgent` and `AssistantAgent` is still `None`, and if it is `None` the code is guaranteed to get an error. So, by this logic the default should be `False`. 

Another thing is the `default_auto_reply` argument, whose default is an empty string but the type hint is still `Optional...`. We should remove the `Optional` bit.

Comment (ekzhu) [reply autogen/agentchat/conversable_agent.py:1]: cc @microsoft/autogen for awareness.

File: autogen/oai/client.py (hunk 2 above)
Comment (gunnarku) [reply autogen/oai/client.py:368]: @ekzhu : Thank you, will update the error message and a corresponding test case.

File: autogen/agentchat/conversable_agent.py
Comment (gunnarku) [reply autogen/agentchat/conversable_agent.py:1]: LMK if you want me to make these changes in this PR as well:

- Default `llm_config` to `False`.
- Remove `Optional` from `default_auto_reply`.

File: autogen/agentchat/assistant_agent.py [hunk 3]
```diff
@@ -33,7 +33,7 @@ def __init__(
         self,
         name: str,
         system_message: Optional[str] = DEFAULT_SYSTEM_MESSAGE,
-        llm_config: Optional[Union[Dict, Literal[False]]] = None,
+        llm_config: Optional[Union[Dict, Literal[False]]] = False,
```
Comment (sonichi) [inline autogen/agentchat/assistant_agent.py:36]: One problem of this change is that, if a user omits `llm_config`, no error will be raised. AssistantAgent is intended to use a non-empty `llm_config`, unless explicitly set to False. So a silent failure won't be good.

File: autogen/agentchat/conversable_agent.py [hunk 4]
```diff
@@ -78,8 +78,8 @@ def __init__(
         human_input_mode: Optional[str] = "TERMINATE",
         function_map: Optional[Dict[str, Callable]] = None,
         code_execution_config: Union[Dict, Literal[False]] = False,
-        llm_config: Optional[Union[Dict, Literal[False]]] = None,
-        default_auto_reply: Optional[Union[str, Dict, None]] = "",
+        llm_config: Optional[Union[Dict, Literal[False]]] = False,
```
Comment (sonichi) [inline autogen/agentchat/conversable_agent.py:81]: Same problem with this change. We probably want to raise the error explicitly when llm_config is omitted for general ConversableAgent.

File: autogen/oai/client.py [hunk 5]
```diff
@@ -345,15 +345,27 @@ def __init__(self, *, config_list: Optional[List[Dict[str, Any]]] = None, **base
...
+            if "model" not in extra_kwargs:
+                raise ValueError("Please specify a value for the 'model' in 'llm_config'.")
 
         self._clients: List[ModelClient] = []
         self._config_list: List[Dict[str, Any]] = []
 
         if config_list:
             config_list = [config.copy() for config in config_list]  # make a copy before modifying
             for config in config_list:
+                # We require that each element of `config_list` has a non-empty value
```
Comment (sonichi) [inline autogen/oai/client.py:362]: When `extra_kwargs` contains "model", we don't require "model" to be contained in each element of `config_list`.

File: autogen/agentchat/conversable_agent.py [hunk 6]
```diff
@@ -143,6 +143,11 @@ def __init__(
             self.llm_config = self.DEFAULT_CONFIG.copy()
             if isinstance(llm_config, dict):
                 self.llm_config.update(llm_config)
+            # We still have a default `llm_config` because the user didn't
+            # specify anything. This won't work, so raise an error to avoid
+            # an obscure message from the OpenAI service.
+            if self.llm_config == self.DEFAULT_CONFIG:
+                raise ValueError("Please specify the value for 'llm_config'.")
```
Comment (sonichi) [inline autogen/agentchat/conversable_agent.py:150]: It's not always necessary to raise. If the DEFAULT_CONFIG is modified by the developer to be a valid llm_config, then we shouldn't raise.

File: autogen/agentchat/assistant_agent.py (hunk 3 above)
Comment (gunnarku) [reply autogen/agentchat/assistant_agent.py:36]: OK, will undo.

File: autogen/agentchat/conversable_agent.py (hunk 4 above)
Comment (gunnarku) [reply autogen/agentchat/conversable_agent.py:81]: OK, will undo.

File: autogen/oai/client.py (hunk 5 above)
Comment (gunnarku) [reply autogen/oai/client.py:362]: OK, will fix.

File: autogen/agentchat/conversable_agent.py (hunk 6 above)
Comment (gunnarku) [reply autogen/agentchat/conversable_agent.py:150]: OK, will explicitly check against `{}` to make sure that there's configuration.

File: autogen/agentchat/assistant_agent.py (hunk 3 above)
Comment (gunnarku) [reply autogen/agentchat/assistant_agent.py:36]: Yes, this is my omission. Thank you for pointing this out! Will fix it shortly.

**Review by ekzhu — APPROVED** [review]

**Pull request merged**